### PR TITLE
Fix $redirectUri sometimes has two '?'

### PR DESCRIPTION
**Pull request by RussellAult**
Resolves Issue #17885 

Check getRedirectUri() for queries, and add a '&' instead of a '?' to $redirectUri if it already has them; otherwise, $redirectUri might end up with two '?'.

### CODE REVIEW

**Review by ChristophWurst — APPROVED** [review]
Makes sense! Thanks! :v:

**Review by juliusknorr — APPROVED** [review]

File: core/Controller/ClientFlowLoginController.php [hunk 1]
```diff
@@ -337,9 +338,16 @@ public function generateAppPassword($stateToken,
...
 
+			if (parse_url($client->getRedirectUri(), PHP_URL_QUERY)) {
+				$redirectUriSeparator = '&';
+			} else {
+				$redirectUriSeparator = '?';
+			}
+
 			$redirectUri = sprintf(
-				'%s?state=%s&code=%s',
+				'%s%sstate=%s&code=%s',
```
Comment (kesselb) [inline core/Controller/ClientFlowLoginController.php]: I think this is hard to read. I would put the $client->getRedirectUri() before the sprintf for better readability. But this is just my personal preference ;)

**Comment by kesselb** [comment]
```
Index: tests/Core/Controller/ClientFlowLoginControllerTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/Core/Controller/ClientFlowLoginControllerTest.php	(revision 4665efeffa02bdbf7a231a23c9c1ed7293bec0c6)
+++ tests/Core/Controller/ClientFlowLoginControllerTest.php	(date 1573504734897)
@@ -398,7 +398,16 @@
 		$this->assertEquals($expected, $this->clientFlowLoginController->generateAppPassword('MyStateToken'));
 	}
 
-	public function testGeneratePasswordWithPasswordForOauthClient() {
+	/**
+	 * @param string $redirectUri
+	 * @param string $redirectUrl
+	 *
+	 * @testWith
+	 * ["https://example.com/redirect.php", "https://example.com/redirect.php?state=MyOauthState&code=MyAccessCode"]
+	 * ["https://example.com/redirect.php?hello=world", "https://example.com/redirect.php?hello=world&state=MyOauthState&code=MyAccessCode"]
+	 *
+	 */
+	public function testGeneratePasswordWithPasswordForOauthClient($redirectUri, $redirectUrl) {
 		$this->session
 			->expects($this->at(0))
 			->method('get')
@@ -471,7 +480,7 @@
 			->willReturn($token);
 		$client = new Client();
 		$client->setName('My OAuth client');
-		$client->setRedirectUri('https://example.com/redirect.php');
+		$client->setRedirectUri($redirectUri);
 		$this->clientMapper
 			->expects($this->once())
 			->method('getByIdentifier')
@@ -481,7 +490,7 @@
 		$this->eventDispatcher->expects($this->once())
 			->method('dispatch');
 
-		$expected = new Http\RedirectResponse('https://example.com/redirect.php?state=MyOauthState&code=MyAccessCode');
+		$expected = new Http\RedirectResponse($redirectUrl);
 		$this->assertEquals($expected, $this->clientFlowLoginController->generateAppPassword('MyStateToken', 'MyClientIdentifier'));
 	}
```

Small testset for your change. Feel free to push it if you have access to your branch. But we can still add this later.

File: core/Controller/ClientFlowLoginController.php (hunk 1 above)
Comment (RussellAult) [reply core/Controller/ClientFlowLoginController.php]: You're not wrong. I'm new to the codebase so I was trying to do the change that would modify existing code the least, but I'm sure there's a DRY-er way to do this.

**Comment by RussellAult** [comment]
Apparently I don't know how to make the DCO check happy. Can't do it through the website; tried doing it through the GitHub Desktop software (which added two more commits but didn't solve the problem). I can close this PR and try again with a fresh branch if that's necessary.

**Comment by gary-kim** [comment]
You need to amend your previous commits then force push then. So something like `git commit -s --amend` then `git push --force`. You may also want to squash your commits.

**Comment by RussellAult** [comment]
@gary-kim I spent the better part of an hour trying (and failing) to figure out how to do that. Basically, I'm not a developer, I don't know Git at all. I run Nextcloud for myself, noticed a bug, and wanted to give back. I treat GitHub like Wikipedia (i.e. it's easy to make small changes using only a browser). I'm not sure what "squashing commits" even means.

In truth, I've now spent more time trying to figure out how to sign-off on already-submitted commits than I did troubleshooting and creating the original PR. I even downloaded GitHub Desktop and followed those git commands, but it only created new commits and didn't sign the old one. If you need the commit signed (even though there's somehow a second commit that is signed with the same contents) I'll close this PR and create a new one (from a new branch), but I'm at my wits end on figuring out Git.

**Comment by gary-kim** [comment]
You can open a new PR if you'd like but I think it'd be easier to keep the current one so we can keep the approvals :).
On your computer open a terminal and navigate to the folder in which you put the repository `cd DIRECTORY`, then run `git checkout patch-1`, then `git rebase -i HEAD~4`. Your default text editor will be opened. From here, change all instances of the word `pick` that's not commented out with `squash` then save and close the file. Another file will be opened that you can edit to change the commit message. Save and close that file then run `git push --force`. 
What this will do is squash all your commits into one then force push to this branch. Your last commit has a DCO already so your one commit already has your sign off.

Git does take a bit to get used to, it's fine.

File: tests/Core/Controller/ClientFlowLoginControllerTest.php [hunk 2]
```diff
@@ -481,7 +490,7 @@ public function testGeneratePasswordWithPasswordForOauthClient() {
 		$this->eventDispatcher->expects($this->once())
 			->method('dispatch');
 
-		$expected = new Http\RedirectResponse('https://example.com/redirect.php?state=MyOauthState&code=MyAccessCode');
+		$expected = new Http\RedirectResponse($redirectUri);
```
Comment (kesselb) [inline tests/Core/Controller/ClientFlowLoginControllerTest.php]:
```suggestion
		$expected = new Http\RedirectResponse($redirectUrl);
```

Than the tests will pass ;)

**Comment by kesselb** [comment]
Thanks @RussellAult :+1: I think we are good without the DCO here @ChristophWurst @juliushaertl? Probably you can merge this pr and squash the commits (if this feature is enabled for this repository)?

**Review by kesselb — CHANGES_REQUESTED** [review]
Looks good but tests fails because of typo

File: tests/Core/Controller/ClientFlowLoginControllerTest.php (hunk 2 above)
Comment (RussellAult) [reply tests/Core/Controller/ClientFlowLoginControllerTest.php]: Whoops, sorry!

**Comment by rullzer** [comment]
I took the liberty of squasing your commits!
Thanks a lot for you work.

**Review by rullzer — APPROVED** [review]

**Review by kesselb — APPROVED** [review]

**Comment by kesselb** [comment]
> I took the liberty of squasing your commits!

How do you do that? ;)

**Comment by rullzer** [comment]
> > I took the liberty of squasing your commits!
> 
> How do you do that? ;)

add his server as a remote.
Checkout, squash etc, push.

**Pull request merged**